src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WrapWithTag/WrapWithTagEndpoint.cs

```diff
@@ -0,0 +1,107 @@
+        private readonly DocumentResolver _documentResolver;
+        private readonly RazorDocumentMappingService _razorDocumentMappingService;
+        private readonly ILogger _logger;
+
+        public WrapWithTagEndpoint(
+            ClientNotifierServiceBase languageServer,
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            RazorDocumentMappingService razorDocumentMappingService,
+            ILoggerFactory loggerFactory)
+        {
+            if (languageServer is null)
+            {
+                throw new ArgumentNullException(nameof(languageServer));
+            }
+
+            if (projectSnapshotManagerDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorDocumentMappingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorDocumentMappingService));
+            }
+
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _languageServer = languageServer;
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+            _documentResolver = documentResolver;
+            _razorDocumentMappingService = razorDocumentMappingService;
+            _logger = loggerFactory.CreateLogger<WrapWithTagEndpoint>();
+        }
+
+        public async Task<WrapWithTagResponse?> Handle(WrapWithTagParams request, CancellationToken cancellationToken)
+        {
+            var documentSnapshot = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                _documentResolver.TryResolveDocument(request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+
+                return documentSnapshot;
+            }, cancellationToken).ConfigureAwait(false);
+
+            if (documentSnapshot is null || cancellationToken.IsCancellationRequested)
+            {
+                _logger.LogWarning($"Failed to find document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            if (!request.Range.Start.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex))
```

Comment:
Ahh if I recall correctly we were doing this elsewhere safe-guard ourselves from Razor compiler bugs. I don't think that applies here meaning I think we can rely fully on the GetLanguageKind below. @ryanbrandenburg might remember differently though

Reply:
We have yet to prove that we've resolved all of the issues which might cause the your requested range to be out of bounds. I think this is fine for now, and once we are more certain that everything has been resolved we can do a pass to remove all "TryGetAbsoluteIndex" calls.